devel/cmake/Portfile

```diff
@@ -104,6 +109,10 @@ if {${subport} eq ${name}} {
         patch-fix_cxx14_17_checks.devel.diff \
         patch-fix-system-prefix-path.devel.diff
 
+    if {${os.major}==17 && (${build_arch} eq "i386" || [variant_isset universal]) && [vercmp ${developerversion} 10] >= 0} {
+        patchfiles-append patch-fix-feature-test.devel.diff
```

Comment:
the same error is also happening to folks running os.major=18 btw, like @Gcenx , so you might make him happy if you also handled that, as he's been asking me to fix this for months

Reply:
of course, that's another corner case, and we can only go so far out of our way to make these corner case folks happy without making ourselves crazy in the process.

Reply:
@kencu do you think that are we better off with these "if" statements for corner cases, or does it hurt to just patch for every install? I can honestly go either way, especially here, since sometimes a massive patch is more useful even if it means we give up some info for when linking has warnings or the like.

Reply:
@kencu if it’s changed to >=17 that would also cover Mojave universal and since base by default disabled universal above Mojave it wouldn’t cause any issues.

But this only covers cmake I’d also encountered issues with other cmake group ports. So I went back to patching base.

Reply:
@michaelld -- I would say if you go with the patchfile to supress the i386 warning, it could be applied always, as it only cuts out that one warning, I believe. 

if you go with a system test, I would do this in the portfile (pseudocode):
```
if Xcode >= 10 and universal
   configure.ldflags-append -Wl,-w
end
```
as that will be resilient forever, whatever Cmake does in the future, whereas the patch will break eventually, someday, and I just hate patching unnecessarily. But that is me.

Regarding  the SDK, I would not change what the portfile does now, as it has survived years and years of trials. I would tell anyone who wants to build universal with Xcode 10+ that they have to install the MacOSX10.13.sdk and then build against it. On 10.13, building against the MacOSX10.13.sdk happens automagically as that is how base is written (as my test above). On 10.14, you have to do it manually, but either way, base already does the right thing and I don't believe we want to go backwards and make macports starting building against "/" at this point in time.

If anyone thinks that asking someone who wants to build universal on 10.13 with Xcode10+ to install the MacOSX10.13.sdk is too much to ask, I would not agree with that, as anyone doing that should darned well know just what they are doing.

Reply:
One thing to ponder -- as @Gcenx points out -- it's possible this same linker warning breaks other universal builds later on that use cmake, in the same way. I haven't yet come across those, but if that were the case (I assume it is, as @Gcenx is not commonly incorrect) and if Marcus' patchfile were to also remove this warning from builds using cmake, not just the build of cmake itself, then his patch would be 100% the right way to go.

Also, if you don't mind watching for patch breakage over the next few years, Marcus' patch is definitely cleaner in that it removes only that one warning.

Reply:
> Regarding the SDK, I would not change what the portfile does now, as it has survived years and years of trials.

I am reasonably sure the SDK code needs to be fixed.
Perhaps I should not have included it in this pull request because it is **independent** of the universal variant.
The problem just came to light when I happen to be messing around with a universal build.

I am sorry if I was not clear before.
The code `configure.env-append CMAKE_OSX_SYSROOT=...` does nothing.
Perhaps it did at some point but  [no longer](https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L45).
If we want to set CMAKE_OSX_SYSROOT (which I would argue we should), we must set the environmental variable `SDKROOT`.
MacPorts mostly takes care of that for us, unless `${configure.sdkroot}` is empty, which is what the new code in the pull request handles.

Please let me know if I am still not being clear.

Reply:
> Marcus' patchfile were to also remove this warning from builds using cmake, not just the build of cmake itself ...

I believe that the patch affects only one small part of the build of cmake and not any ports that use cmake.

Reply:
@MarcusCalhoun-Lopez CMAKE_OSX_SYSROOT doesn’t need to be set when building cmake only when using cmake to build something.

Reply:
If we do not set CMAKE_OSX_SYSROOT, then cmake searches for its own.
This is definitely not what we want.
Building any port should, as much as possible, respect the MacPorts configuration.

Reply:
But the cmake portfile already sets this correctly, to ${configure.sdkroot}, which is perfect.

What am I missing?

Reply:
I think the error Marcus saw in building cmake +universal was caused by base setting the SDK to MacOSX.sdk, which is the fallback it will go to if MacOSX10.13.sdk doesn't exist (on 10.13). The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.

I feel like I must be looking at something sideways here.

Reply:
Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot so the change isn’t required, cmake “should” respect sdkroot but it might no in some cases? That’s lightly why the original Portfile was setting CMAKE_OSX_SYSROOT.

That would also ensure cmake doesn’t try overwriting sdkroot with the sdk it selected

Reply:
Please forgive my inability to properly convey the issue.
Also, please forgive me if, as a result, my explanation is overly verbose.
1. The current Portfile code (`configure.env-append CMAKE_OSX_SYSROOT=...`) does **nothing**&nbsp;[1]. At the very least we should remove it.
1. When building cmake, cmake attempts to find a value for `CMAKE_OSX_SYSROOT`. It guesses as best it can&nbsp;[1].
1. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags&nbsp;[2].
1. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.

> Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot

> But the cmake portfile already sets this correctly

I do not believe this is correct.
I was unable to get this behavior, I could not any reference to `$ENV{CMAKE_OSX_SYSROOT}`, and the there is no indication of this behavior in the cmake code&nbsp;[1].

> cmake to not select the sdk would cause problems if called outside of macports

Projects outside of MacPorts will be unaffected by setting `SDKROOT` during the **build** of cmake.
Even if they were, the base sets `SDKROOT` anyway.

> The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.

We can certainly modify Xcode and command line tools until we find a way to have cmake guess the right value for `CMAKE_OSX_SYSROOT`, but it makes much more sense to set it to the value we want in the first place.

Please let me know if I am still not explaining the situation correctly.

[1] https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L42
[2] https://cmake.org/cmake/help/v3.15/variable/CMAKE_OSX_SYSROOT.html

Reply:
> 1. The current Portfile code (`configure.env-append CMAKE_OSX_SYSROOT=...`) does **nothing** [1]. At the very least we should remove it.


```
    if {${configure.sdkroot} != ""} {
        configure.env-append CMAKE_OSX_SYSROOT=${configure.sdkroot}
    } else {
        configure.env-append CMAKE_OSX_SYSROOT=/
    }
```

The code does nothing unless macports-base sets an SDK.

If sdkroot is set in my case I'm forcing the 10.13SDK on Mojave, then I will get;
`CMAKE_OSX_SYSROOT=${configure.sdkroot}`
As sdkroot set your getting
`configure.env-append CMAKE_OSX_SYSROOT=/`
The issue is macports-base isn't setting an SDK in your case.

> 2. When building cmake, cmake attempts to find a value for `CMAKE_OSX_SYSROOT`. It guesses as best it can [1].

If that is not set then yes cmake will figure it out on it's own.
```cmake
if(CMAKE_OSX_SYSROOT)
  # Use the existing value without further computation to choose a default.
  set(_CMAKE_OSX_SYSROOT_DEFAULT "${CMAKE_OSX_SYSROOT}")
````

If not it will attempt to use `SDKROOT` to set `CMAKE_OSX_SYSROOT`
```cmake
elseif(NOT "x$ENV{SDKROOT}" STREQUAL "x" AND
        (NOT "x$ENV{SDKROOT}" MATCHES "/" OR IS_DIRECTORY "$ENV{SDKROOT}"))
  # Use the value of SDKROOT from the environment.
  set(_CMAKE_OSX_SYSROOT_DEFAULT "$ENV{SDKROOT}")
```

> 3. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags [2].

Seems normal for cmake that's cmake's own way of passing the SDK onto clang/gcc but clang can also just use SDKROOK if -isysroot isn't set on recent clang .

> 4. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.

You were talking about patching that out that check before, that would be bad.

> 
> > Setting CMAKE_OSX_SYSROOT to configure.sdk also sets sdkroot
> 
> > But the cmake portfile already sets this correctly
> 
> I do not believe this is correct.
> I was unable to get this behavior, I could not any reference to `$ENV{CMAKE_OSX_SYSROOT}`, and the there is no indication of this behavior in the cmake code [1].

Read my comment above.


> 
> > cmake to not select the sdk would cause problems if called outside of macports
> 
> Projects outside of MacPorts will be unaffected by setting `SDKROOT` during the **build** of cmake.
> Even if they were, the base sets `SDKROOT` anyway.

You were talking about patching that out that check before.


> > The proper fix for that, IMHO, is to install the MacOSX10.13.sdk.
> 
> We can certainly modify Xcode and command line tools until we find a way to have cmake guess the right value for `CMAKE_OSX_SYSROOT`, but it makes much more sense to set it to the value we want in the first place.

Have you tried placing the 10.13SDK into the location @kencu mention above?
> 
> Please let me know if I am still not explaining the situation correctly.
> 
> [1] https://gitlab.kitware.com/cmake/cmake/blob/master/Modules/Platform/Darwin-Initialize.cmake#L42
> [2] https://cmake.org/cmake/help/v3.15/variable/CMAKE_OSX_SYSROOT.html

Reply:
> ```cmake
> if(CMAKE_OSX_SYSROOT)
>   # Use the existing value without further computation to choose a default.
>   set(_CMAKE_OSX_SYSROOT_DEFAULT "${CMAKE_OSX_SYSROOT}")
> ```
> 
> If not it will attempt to use `SDKROOT` to set `CMAKE_OSX_SYSROOT`

Yes, I agree.
In fact, that the point I am (poorly) trying to make.
`CMAKE_OSX_SYSROOT` is not defined in the Portfile, no matter the system configuration.
`ENV{CMAKE_OSX_SYSROOT}` is [defined](https://cmake.org/cmake/help/latest/variable/ENV.html), but `CMAKE_OSX_SYSROOT` is not.
The Portfile sets the environmental variable `CMAKE_OSX_SYSROOT`, which is **not** the same thing as passing `-DCMAKE_OSX_SYSROOT=...` to cmake.
That is why `configure.env-append CMAKE_OSX_SYSROOT=...` does not do anything.

Please correct me if I am wrong here.

> > 1. When building cmake, cmake adds `-isysroot ${CMAKE_OSX_SYSROOT}` to the compiler flags [2].
> 
> Seems normal for cmake that's cmake's own way of passing the SDK onto clang/gcc but clang can also just use SDKROOK if -isysroot isn't set on recent clang .

Yes, but cmake **is** setting `-isysroot`, sometimes to the wrong value.

> 
> > 1. The value of `CMAKE_OSX_SYSROOT`, whatever it is, is **not** recorded in the installation. In other words, no port that uses cmake is affected.
> 
> You were talking about patching that out that check before, that would be bad.

Tha patch is for a completely unrelated issue.
I should have opened two different pull requests.
Sorry for the confusion.
The patch I proposed has no affect on `SDKROOT`, `CMAKE_OSX_SYSROOT`, `-isysroot`, or any ports that use cmake.

> Have you tried placing the 10.13SDK into the location @kencu mention above?

Yes, I have.
However, instead of modifying system files, it makes much more sense to simply set `CMAKE_OSX_SYSROOT` to the value we want, as the Portfile currently **tries** to do.

Reply:
> ENV{CMAKE_OSX_SYSROOT} is defined, but CMAKE_OSX_SYSROOT is not.

Now I see your point. I'm sorry I'm so slow. I assumed that the cmake build picked up CMAKE_OSX_SYSROOT from the environment (otherwise why set it in the Portfile?) but I did not go back and challenge the original premise that it did. If it truly does not, then you're right, that whole part of the portfile that sets CMAKE_OSX_SYSROOT in the environment is pointless.


> Have you tried placing the 10.13SDK into the location @kencu mention above?
> 
> Yes, I have.
> However, instead of modifying system files, it makes much more sense to simply set CMAKE_OSX_SYSROOT to the value we want, as the Portfile currently tries to do.

You might tweak cmake to build universal without the MacOSX10.13.sdk in the proper place, but nothing else will build universal unless likewise hacked. It's trivially easy to put the SDK there, and I can't help but believe it's just the right thing to do.